packages/edit-site/src/components/left-sidebar/navigation-panel/template-switcher/index.js

```diff
@@ -0,0 +1,213 @@
+	__experimentalNavigationItem as NavigationItem,
+} from '@wordpress/components';
+import { Icon, home, plus, undo } from '@wordpress/icons';
+
+/**
+ * Internal dependencies
+ */
+import TemplatePreview from './template-preview';
+import ThemePreview from './theme-preview';
+
+const TEMPLATE_OVERRIDES = {
+	page: ( slug ) => `page-${ slug }`,
+	category: ( slug ) => `category-${ slug }`,
+	post: ( slug ) => `single-post-${ slug }`,
+};
+
+function TemplateLabel( { template, homeId } ) {
+	return (
+		<>
+			{ template.slug }{ ' ' }
+			{ template.id === homeId && (
+				<Tooltip text={ __( 'Home' ) }>
+					<div className="edit-site-template-switcher__label-home-icon">
+						<Icon icon={ home } />
+					</div>
+				</Tooltip>
+			) }
+			{ template.status !== 'auto-draft' && (
+				<Tooltip text={ __( 'Customized' ) }>
+					<span className="edit-site-template-switcher__label-customized-dot" />
+				</Tooltip>
+			) }
+		</>
+	);
+}
+
+function NavigationItemWithIcon( { icon, title, ...props } ) {
+	return (
+		<NavigationItem
+			title={
+				<>
+					{ title }{ ' ' }
+					{ icon && (
+						<div className="edit-site-template-switcher__label-home-icon">
+							<Icon icon={ icon } />
+						</div>
+					) }
+				</>
+			}
+			{ ...props }
+		/>
+	);
+}
+
+export default function TemplateSwitcher( {
+	page,
+	activeId,
+	onActiveIdChange,
+	onActiveTemplatePartIdChange,
+	onAddTemplate,
+	onRemoveTemplate,
+} ) {
+	const [ hoveredTemplate, setHoveredTemplate ] = useState();
+	const [ themePreviewVisible, setThemePreviewVisible ] = useState( false );
+
+	const onHoverTemplatePart = ( id ) => {
+		setHoveredTemplate( { id, type: 'template-part' } );
+	};
+	const onMouseEnterTheme = () => {
+		setThemePreviewVisible( () => true );
+	};
+	const onMouseLeaveTheme = () => {
+		setThemePreviewVisible( () => false );
+	};
+
+	const { currentTheme, template, templateParts, homeId } = useSelect(
+		( select ) => {
+			const {
+				getCurrentTheme,
+				getEntityRecord,
+				getEntityRecords,
+			} = select( 'core' );
+
+			const _template = getEntityRecord(
+				'postType',
+				'wp_template',
+				activeId
+			);
+
+			const { getHomeTemplateId } = select( 'core/edit-site' );
+
+			return {
+				currentTheme: getCurrentTheme(),
+				template: _template,
+				templateParts: _template
+					? getEntityRecords( 'postType', 'wp_template_part', {
+							resolved: true,
+							template: _template.slug,
+					  } )
+					: null,
+				homeId: getHomeTemplateId(),
+			};
+		},
+		[ activeId ]
+	);
+
+	const templateItem = {
+		label: template ? (
+			<TemplateLabel template={ template } homeId={ homeId } />
+		) : (
+			__( 'Loading…' )
+		),
+		value: activeId,
+		slug: template ? template.slug : __( 'Loading…' ),
+		content: template?.content,
+	};
+
+	const templatePartItems = templateParts?.map( ( templatePart ) => ( {
+		label: <TemplateLabel template={ templatePart } />,
+		value: templatePart.id,
+		slug: templatePart.slug,
+	} ) );
```

Comment:
We probably don't need this map here, as we can just as easily do it once in the render below.
E.g.
```jsx
{ templateParts?.map( ( templatePart ) => (
	<NavigationItem
		key={ `template-part-${ templatePart.id }` }
		item={ templatePart.slug }
		title={ <TemplateLabel template={ templatePart } /> }
		onClick={ () => onActiveTemplatePartIdChange( templatePart.id ) }
		onMouseEnter={ () => onHoverTemplatePart( templatePart.id ) }
		onMouseLeave={ () => onHoverTemplatePart( null ) }
	/>
) ) }
```

Note also that while there are no checks, `title` is supposed to be a string.
I guess that if we [unify `TemplateLabel` and `NavigationItemWithIcon`](https://github.com/WordPress/gutenberg/pull/25615/files#r494383927) this inconsistency wouldn't happen anymore.

Reply:
🤔 If we are supposed to use `string` as `title`, then I need to copy-paste the whole part from the `NavigationItem` to get the same styling with a custom icon.